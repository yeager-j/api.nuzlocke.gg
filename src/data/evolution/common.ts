/**
 * Core evolution triggers that can cause a Pokemon to evolve
 */
export enum EvolutionTrigger {
  LEVEL_UP = "LEVEL_UP",
  TRADE = "TRADE",
  ITEM_USE = "ITEM_USE",
  SPECIAL = "SPECIAL", // For unique evolution mechanics
}

/**
 * Time of day requirements for evolution
 */
export enum TimeOfDay {
  ANY = "ANY",
  DAY = "DAY",
  NIGHT = "NIGHT",
  DUSK = "DUSK", // Special case for Rockruff -> Lycanroc Dusk Form
}

/**
 * Weather conditions that can affect evolution
 */
export enum Weather {
  ANY = "ANY",
  RAIN = "RAIN",
  THUNDERSTORM = "THUNDERSTORM",
  SNOW = "SNOW",
  HARSH_SUNLIGHT = "HARSH_SUNLIGHT",
  SANDSTORM = "SANDSTORM",
  FOG = "FOG",
}

/**
 * Gender-specific evolution requirements
 */
export enum Gender {
  ANY = "ANY",
  MALE = "MALE",
  FEMALE = "FEMALE",
}
