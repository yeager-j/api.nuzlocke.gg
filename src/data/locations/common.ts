import { KantoLocationData } from "@/data/locations/regions/kanto/locations";

export enum EncounterMethod {
  TALL_GRASS = "TALL_GRASS",
  CAVE = "CAVE",
  FISHING = "FISHING",
  SURFING = "SURFING",
  INTERACT = "INTERACT",
  TRADE = "TRADE",
  GIFT = "GIFT",
  EVOLVE = "EVOLVE",
}

export const EncounterMethodNames = {
  [EncounterMethod.TALL_GRASS]: "Tall Grass",
  [EncounterMethod.CAVE]: "Cave",
  [EncounterMethod.FISHING]: "Fishing",
  [EncounterMethod.SURFING]: "Surfing",
  [EncounterMethod.INTERACT]: "Interact",
  [EncounterMethod.TRADE]: "Trade",
  [EncounterMethod.GIFT]: "Gift",
  [EncounterMethod.EVOLVE]: "Evolution",
};

export const PokemonLocationData = {
  ...KantoLocationData,
};
