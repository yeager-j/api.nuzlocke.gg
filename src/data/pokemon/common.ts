import { PokemonVersionGroups } from "@/data/versions";

export const availableStartingUSUM: PokemonVersionGroups[] = [
  PokemonVersionGroups.ULTRA_SUN_ULTRA_MOON,
];

export const availableStartingSM: PokemonVersionGroups[] = [
  ...availableStartingUSUM,
  PokemonVersionGroups.SUN_MOON,
];

export const availableStartingORAS: PokemonVersionGroups[] = [
  ...availableStartingSM,
  PokemonVersionGroups.OMEGA_RUBY_ALPHA_SAPPHIRE,
];

export const availableStartingXY: PokemonVersionGroups[] = [
  ...availableStartingORAS,
  PokemonVersionGroups.XY,
];

export const availableStartingB2W2: PokemonVersionGroups[] = [
  ...availableStartingXY,
  PokemonVersionGroups.BLACK_2_WHITE_2,
];

export const availableStartingBW: PokemonVersionGroups[] = [
  ...availableStartingB2W2,
  PokemonVersionGroups.BLACK_WHITE,
];

export const availableStartingHGSS: PokemonVersionGroups[] = [
  ...availableStartingBW,
  PokemonVersionGroups.HEARTGOLD_SOULSILVER,
];

export const availableStartingPt: PokemonVersionGroups[] = [
  ...availableStartingHGSS,
  PokemonVersionGroups.PLATINUM,
];

export const availableStartingDP: PokemonVersionGroups[] = [
  ...availableStartingPt,
  PokemonVersionGroups.DIAMOND_PEARL,
];

export const availableStartingE: PokemonVersionGroups[] = [
  ...availableStartingDP,
  PokemonVersionGroups.EMERALD,
];

export const availableStartingFRLG: PokemonVersionGroups[] = [
  ...availableStartingE,
  PokemonVersionGroups.FIRERED_LEAFGREEN,
];

export const availableStartingRS: PokemonVersionGroups[] = [
  ...availableStartingFRLG,
  PokemonVersionGroups.RUBY_SAPPHIRE,
];

export const availableStartingC: PokemonVersionGroups[] = [
  ...availableStartingRS,
  PokemonVersionGroups.CRYSTAL,
];

export const availableStartingGS: PokemonVersionGroups[] = [
  ...availableStartingC,
  PokemonVersionGroups.GOLD_SILVER,
];

export const availableStartingY: PokemonVersionGroups[] = [
  ...availableStartingGS,
  PokemonVersionGroups.YELLOW,
];

export const availableStartingRB: PokemonVersionGroups[] = [
  ...availableStartingY,
  PokemonVersionGroups.RED_BLUE,
];

export const versionsWithoutHiddenAbilities: PokemonVersionGroups[] = [
  PokemonVersionGroups.RUBY_SAPPHIRE,
  PokemonVersionGroups.FIRERED_LEAFGREEN,
  PokemonVersionGroups.EMERALD,
  PokemonVersionGroups.DIAMOND_PEARL,
  PokemonVersionGroups.PLATINUM,
  PokemonVersionGroups.HEARTGOLD_SOULSILVER,
];

export const gen1Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.RED_BLUE,
  PokemonVersionGroups.YELLOW,
];

export const gen2Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.GOLD_SILVER,
  PokemonVersionGroups.CRYSTAL,
];

export const gen3Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.RUBY_SAPPHIRE,
  PokemonVersionGroups.FIRERED_LEAFGREEN,
  PokemonVersionGroups.EMERALD,
];

export const gen4Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.DIAMOND_PEARL,
  PokemonVersionGroups.PLATINUM,
  PokemonVersionGroups.HEARTGOLD_SOULSILVER,
];

export const gen5Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.BLACK_WHITE,
  PokemonVersionGroups.BLACK_2_WHITE_2,
];

export const gen6Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.XY,
  PokemonVersionGroups.OMEGA_RUBY_ALPHA_SAPPHIRE,
];

export const gen7Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.SUN_MOON,
  PokemonVersionGroups.ULTRA_SUN_ULTRA_MOON,
];

export const gen8Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.SWORD_SHIELD,
];

export const gen9Games: PokemonVersionGroups[] = [
  PokemonVersionGroups.SCARLET_VIOLET,
];
