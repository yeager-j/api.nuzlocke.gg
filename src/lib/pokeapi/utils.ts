import { PokemonVersionGroup } from "@/data/versions";

export function toPascalCase(str: string) {
  const regexp = str.match(
    /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g,
  );

  return (regexp ?? [])
    .map((x) => x.charAt(0).toUpperCase() + x.slice(1).toLowerCase())
    .join("");
}

export function arraysEqual(
  a: PokemonVersionGroup[],
  b: PokemonVersionGroup[],
): boolean {
  if (a.length !== b.length) return false;
  return (
    a.every((item) => b.includes(item)) && b.every((item) => a.includes(item))
  );
}

export function isSubset(
  subset: PokemonVersionGroup[],
  superset: PokemonVersionGroup[],
): boolean {
  return subset.every((item) => superset.includes(item));
}
