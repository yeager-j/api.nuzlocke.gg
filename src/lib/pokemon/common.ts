import path from "path";

export const POKEMON_OUTPUT_PATH = path.join(
  process.cwd(),
  "public/v1/pokemon",
);

export const POKEMON_DATA_PATH = path.join(
  process.cwd(),
  "src/data/pokemon/master",
);
