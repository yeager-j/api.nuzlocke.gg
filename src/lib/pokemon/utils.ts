import {
  Pokemon,
  PokemonMode,
  PokemonSpecies,
  VersionedPokemon,
  VersionedPokemonMode,
  VersionedPokemonSpecies,
} from "@/data/pokemon/types";
import { VersionedProperty } from "@/data/types";
import { PokemonVersionGroup } from "@/data/versions";

/**
 * Retrieves the data corresponding to the specified version from the provided
 * versioned property. If no data matches the given version, the default value
 * is returned.
 *
 * @param {VersionedProperty<T>} data - The versioned property containing data for different versions.
 * @param {PokemonVersionGroup} version - The specific version group to retrieve data for.
 * @return {T} - The data corresponding to the specified version, or the default value if no match is found.
 */
export function getVersionedData<T>(
  data: VersionedProperty<T>,
  version: PokemonVersionGroup,
): T {
  if (data.versions.filter((v) => v.appliesTo.includes(version)).length > 1) {
    console.warn(
      "Multiple versions match the specified version group:",
      version,
    );
  }

  const versionedData = data.versions.find((v) =>
    v.appliesTo.includes(version),
  );

  if (!versionedData) {
    return data.default;
  }

  return versionedData.value;
}

/**
 * Applies the specified version to the provided Pokemon mode, adjusting its types accordingly.
 *
 * @param {PokemonMode} pokemonMode - The Pokemon mode object containing the current configuration of the Pokemon.
 * @param {PokemonVersionGroup} version - The version group to apply to the Pokemon mode, determining version-specific adjustments.
 * @return {VersionedPokemonMode} A new Pokemon mode object with types adapted to the specified version.
 */
export function applyVersionToPokemonMode(
  pokemonMode: PokemonMode,
  version: PokemonVersionGroup,
): VersionedPokemonMode {
  return {
    name: pokemonMode.name,
    displayName: pokemonMode.displayName,
    sprite: pokemonMode.sprite,
    isDefault: pokemonMode.isDefault,
    types: getVersionedData(pokemonMode.types, version),
  };
}

/**
 * Applies a specific version group to a Pokémon form and updates its properties accordingly.
 *
 * @param {Pokemon} pokemonForm - The Pokémon form object to which the version group changes will be applied.
 * @param {PokemonVersionGroup} version - The specific version group used to modify the Pokémon form's data.
 * @return {VersionedPokemon} A new Pokémon object with the version group applied, containing updated evolutionary branches.
 */
export function applyVersionToPokemon(
  pokemonForm: Pokemon,
  version: PokemonVersionGroup,
): VersionedPokemon {
  return {
    name: pokemonForm.name,
    displayName: pokemonForm.displayName,
    modes: pokemonForm.modes
      .filter((mode) => mode.availableIn.includes(version))
      .map((mode) => applyVersionToPokemonMode(mode, version)),
    evolution: {
      ...pokemonForm.evolution,
      evolutionBranches: getVersionedData(
        pokemonForm.evolution.evolutionBranches,
        version,
      ),
    },
  };
}

export function applyVersionToSpecies(
  pokemonSpecies: PokemonSpecies,
  _version: PokemonVersionGroup,
): VersionedPokemonSpecies {
  return {
    name: pokemonSpecies.name,
    displayName: pokemonSpecies.displayName,
    nationalDexNumber: pokemonSpecies.nationalDexNumber,
  };
}
