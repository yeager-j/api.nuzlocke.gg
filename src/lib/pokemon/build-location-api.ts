import fs from "fs/promises";
import path from "path";

import { PokemonLocationData } from "@/data/locations/common";
import { PokemonLocation } from "@/data/locations/types";
import { PokemonDefinition } from "@/data/pokemon/types";
import { PokemonGameMetadata } from "@/data/types";
import {
  PokemonGame,
  PokemonGameData,
  PokemonVersionGroup,
  PokemonVersionGroupData,
} from "@/data/versions";
import { LOCATION_OUTPUT_PATH, POKEMON_DATA_PATH } from "@/lib/pokemon/common";
import {
  LocationOutput,
  PokemonEncounter,
  PokemonOutput,
} from "@/lib/pokemon/types";
import {
  applyVersionToPokemon,
  applyVersionToSpecies,
  loadPokemon,
} from "@/lib/pokemon/utils";

async function ensureDirectories(
  versionId: string,
  gameId: string,
): Promise<void> {
  await fs.mkdir(path.join(LOCATION_OUTPUT_PATH, versionId, gameId), {
    recursive: true,
  });
}

export async function buildLocationApi(): Promise<void> {
  console.log("Starting location JSON generation...");

  try {
    const generations = await fs.readdir(POKEMON_DATA_PATH);
    const pokemonDefinitions: PokemonDefinition[] = [];

    for (const generation of generations) {
      console.log("Collecting Pokemon data from generation: ", generation);
      const pokemonFiles = await fs.readdir(
        path.join(POKEMON_DATA_PATH, generation),
      );

      const pokemonPromises = pokemonFiles.map(async (pokemonFile) => {
        return loadPokemon(generation, pokemonFile);
      });

      const pokemonData = await Promise.all(pokemonPromises);
      pokemonDefinitions.push(...pokemonData);
    }

    for (const version of Object.values(PokemonVersionGroup)) {
      console.log("Processing version: ", version);

      const versionData = PokemonVersionGroupData[version];
      const encounterMap: Map<PokemonLocation, PokemonEncounter[]> = new Map();

      for (const definition of pokemonDefinitions) {
        for (const pokemon of definition.forms) {
          const versionedPokemon = applyVersionToPokemon(pokemon, version);
          const versionedSpecies = applyVersionToSpecies(definition.species);

          const pokemonOutput: PokemonOutput = {
            ...versionedPokemon,
            species: versionedSpecies,
          };

          versionedPokemon.encounters
            .filter((e) => e.method !== "EVOLVE")
            .forEach((e) => {
              const encounters = encounterMap.get(e.location) || [];
              encounters.push({
                pokemon: pokemonOutput,
                method: e.method,
                exclusiveTo: e.exclusiveTo,
              });
              encounterMap.set(e.location, encounters);
            });
        }
      }

      const gamesInVersionGroup = Object.entries(PokemonGameData).filter(
        ([, gameData]) => gameData.versionGroup.includes(version),
      ) as [PokemonGame, PokemonGameMetadata][];

      for (const [game, gameData] of gamesInVersionGroup) {
        const locations: LocationOutput[] = [];

        for (const [location, encounters] of encounterMap) {
          locations.push({
            id: PokemonLocationData[location].id,
            encounters: encounters.filter(
              (e) => !e.exclusiveTo || e.exclusiveTo.includes(game),
            ),
          });
        }

        for (const location of locations) {
          await ensureDirectories(versionData.id, gameData.id);
          await fs.writeFile(
            path.join(
              LOCATION_OUTPUT_PATH,
              versionData.id,
              gameData.id,
              `${location.id}.json`,
            ),
            JSON.stringify(location, null, 4),
          );
        }
      }
    }

    console.log("Location JSON generation complete!");
  } catch (error) {
    console.error("Error generating location JSON:", error);
    throw error;
  }
}
