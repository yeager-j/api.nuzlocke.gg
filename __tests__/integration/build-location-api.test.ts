import path from "path";
import { vol } from "memfs";
import { beforeEach, describe, expect, test, vi } from "vitest";

import mockCaterpie from "../__fixtures__/mock-caterpie";

import { buildLocationApi } from "@/lib/pokemon/build-location-api";
import { LOCATION_OUTPUT_PATH, POKEMON_DATA_PATH } from "@/lib/pokemon/common";
import { loadPokemon } from "@/lib/pokemon/utils";

vi.mock("fs/promises");
vi.mock("@/lib/pokemon/utils", async () => {
  const actual = await vi.importActual("@/lib/pokemon/utils");
  return {
    ...actual,
    loadPokemon: vi.fn(),
  };
});

describe("buildLocationApi integration", () => {
  beforeEach(async () => {
    vol.reset();

    vol.mkdirSync(path.join(POKEMON_DATA_PATH, "generation-1"), {
      recursive: true,
    });

    vol.writeFileSync(
      path.join(POKEMON_DATA_PATH, "generation-1", "10-caterpie.ts"),
      "",
    );

    vi.mocked(loadPokemon).mockImplementation(async (_, pokemonFile) => {
      if (pokemonFile === "10-caterpie.ts") return mockCaterpie;
      throw new Error(`Unexpected pokemon file: ${pokemonFile}`);
    });
  });

  test("correct file structure should exist", async () => {
    await buildLocationApi();

    expect(vol.existsSync(path.join(LOCATION_OUTPUT_PATH, "red"))).toBe(true);
    expect(vol.existsSync(path.join(LOCATION_OUTPUT_PATH, "blue"))).toBe(true);
    expect(vol.existsSync(path.join(LOCATION_OUTPUT_PATH, "firered"))).toBe(
      true,
    );
    expect(vol.existsSync(path.join(LOCATION_OUTPUT_PATH, "leafgreen"))).toBe(
      true,
    );
  });

  test("should build the location files", async () => {
    await buildLocationApi();

    const blueRoute2Json = vol.readFileSync(
      path.join(LOCATION_OUTPUT_PATH, "blue", "route-2.json"),
      "utf8",
    );
    const redRoute2Json = vol.readFileSync(
      path.join(LOCATION_OUTPUT_PATH, "red", "route-2.json"),
      "utf8",
    );

    const fireredViridianForestJson = vol.readFileSync(
      path.join(LOCATION_OUTPUT_PATH, "firered", "viridian-forest.json"),
      "utf8",
    );
    const leafgreenViridianForestJson = vol.readFileSync(
      path.join(LOCATION_OUTPUT_PATH, "leafgreen", "viridian-forest.json"),
      "utf8",
    );

    expect(JSON.parse(blueRoute2Json.toString())).toMatchSnapshot();
    expect(JSON.parse(redRoute2Json.toString())).toMatchSnapshot();
    expect(JSON.parse(fireredViridianForestJson.toString())).toMatchSnapshot();
    expect(
      JSON.parse(leafgreenViridianForestJson.toString()),
    ).toMatchSnapshot();
  });
});
