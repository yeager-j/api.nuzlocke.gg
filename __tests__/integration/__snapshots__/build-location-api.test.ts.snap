// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildLocationApi integration > should build the location files 1`] = `
{
  "encounters": [
    {
      "exclusiveTo": "BLUE",
      "method": "TALL_GRASS",
      "pokemon": {
        "displayName": "Caterpie",
        "evolution": {
          "evolutionBranches": [
            {
              "conditions": [
                {
                  "minLevel": 7,
                  "type": "LEVEL",
                },
              ],
              "to": "metapod",
              "trigger": "LEVEL_UP",
            },
          ],
          "from": "",
        },
        "isDefault": true,
        "name": "caterpie",
        "nationalDexNumber": 10,
        "sprite": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/10.png",
        "types": [
          "Bug",
        ],
      },
    },
  ],
  "id": "route-2",
  "name": "Route 2",
}
`;

exports[`buildLocationApi integration > should build the location files 2`] = `
{
  "encounters": [],
  "id": "route-2",
  "name": "Route 2",
}
`;

exports[`buildLocationApi integration > should build the location files 3`] = `
{
  "encounters": [
    {
      "method": "TALL_GRASS",
      "pokemon": {
        "displayName": "Caterpie",
        "evolution": {
          "evolutionBranches": [
            {
              "conditions": [
                {
                  "minLevel": 7,
                  "type": "LEVEL",
                },
              ],
              "to": "metapod",
              "trigger": "LEVEL_UP",
            },
          ],
          "from": "",
        },
        "isDefault": true,
        "name": "caterpie",
        "nationalDexNumber": 10,
        "sprite": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/10.png",
        "types": [
          "Bug",
        ],
      },
    },
  ],
  "id": "viridian-forest",
  "name": "Viridian Forest",
}
`;

exports[`buildLocationApi integration > should build the location files 4`] = `
{
  "encounters": [
    {
      "method": "TALL_GRASS",
      "pokemon": {
        "displayName": "Caterpie",
        "evolution": {
          "evolutionBranches": [
            {
              "conditions": [
                {
                  "minLevel": 7,
                  "type": "LEVEL",
                },
              ],
              "to": "metapod",
              "trigger": "LEVEL_UP",
            },
          ],
          "from": "",
        },
        "isDefault": true,
        "name": "caterpie",
        "nationalDexNumber": 10,
        "sprite": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/10.png",
        "types": [
          "Bug",
        ],
      },
    },
  ],
  "id": "viridian-forest",
  "name": "Viridian Forest",
}
`;
